README.md
# Sloki
A NodeJS Server for [LokiJS](http://lokijs.org/)

#### :hammer: :hammer: :hammer: :hammer: :hammer: WORK IN PROGRESS :hammer: :hammer: :hammer: :hammer: :hammer:

[![alt CI-badge](https://travis-ci.org/sloki-project/sloki.svg?branch=master)](https://travis-ci.org/sloki-project/sloki)
[![npm version](https://badge.fury.io/js/sloki.svg?v=0)](http://badge.fury.io/js/sloki)
[![Known Vulnerabilities](https://snyk.io/test/github/sloki-project/sloki/badge.svg)](https://snyk.io/test/github/sloki-project/sloki)
[![Dependencies](https://david-dm.org/sloki-project/sloki.svg)](https://david-dm.org/sloki-project/sloki)
[![Dev Dependencies](https://david-dm.org/sloki-project/sloki/dev-status.svg)](https://david-dm.org/sloki-project/sloki?type=dev)

-----
## Documentation
1. [Introduction](#introduction)
    1. [Goal](#Goal)
    2. [Transports](#transports)
    3. [Protocols](#protocols)
        1. [Binary](#binary) (default)
        2. [JSONRPC](#jsonrpc)
        3. [Dinary](#dinary) (default)
2. [Server](#server)
    1. [Installation](#installation)
    2. [Run](#run)
    3. [Help](#help)
3. [Clients](#clients)
    1. [NodeJS](#nodejs)
4. [Benchmarks](#benchmarks)
5. [Development Status](#development-status)
-----

## Introduction

Sloki is a nodejs server which embed [LokiJS](http://lokijs.org/), a blazing fast in-memory documents database.
Sloki help to make LokiJS ***scalable*** : you can now have multiple processes speaking with LokiJS through Sloki.

A possible architecture using sloki :
```                  

    +----------------------------+    TCP / Binary     +-----------------------------------+
    |   NodeJS app worker #1     |<------------------->|                                   |
    +----------------------------+                     |               Sloki               |
                                                       |                                   |
    +----------------------------+    TCP / Dinary     |    +-------------------------+    |
    |  NodeJS app worker #2      |<------------------->|    |                         |    |
    +----------------------------+                     |    |                         |    |
                                                       |    |         LokiJS          |    |
    +----------------------------+    TCP / JSONRPC    |    | fast in-memory database |    |
    | go/php/python/C/whatever   |<------------------->|    |                         |    |
    +----------------------------+                     |    |                         |    |
                                                       |    |                         |    |
    +----------------------------+    TCP / Binary     |    +-------------------------+    |
    |        sloki-cli           |<------------------->|                                   |
    +----------------------------+                     +-----------------------------------+

```
## Goal

The goal is to have a complete in-memory database, supporting documents (mongodb like) AND key/value stores (like redis like).

For documents, [lokiJS](http://lokijs.org) is used.

Key/value stores are not implemented at the moment.


## Transports

For moment, only TCP transport is implemented. The advantage of TCP vs HTTP API is that the connection is persistent (i.e more fast).
Websockets should be implemented before HTTP API.

By default, Sloki listens on the following ports:

| Port      | Transport  | TLS  | Protocol              
|:---------:|------------|------|-----------------
| 6370      | TCP        | NO   | Binary           
| 6371      | TCP        | YES  | Binary
| 6372      | TCP        | NO   | JSONRPC          
| 6373      | TCP        | YES  | JSONRPC
| 6374      | TCP        | NO   | Dinary
| 6375      | TCP        | YES  | Dinary (fastest)

If somebody have an idea why TLS over TCP is fastest than raw TCP, i'd like to know .. :)

You will need a [client](#clients) to speak with sloki.

## Protocols

### **Binary**

The binary protocol has been made with performance in mind. Payloads looks like JSONRPC, but it's not.
```
REQUEST                                     | RESPONSE
------------------------------------------- | --------------------------------------
{                                           | {
    "m":"myMethod",                         |   "r":true,
    "p":["foo","bar"],                      |   "id":"operation-uniq-id"
    "id":"operation-uniq-id"                | }
}                                           |
```
* Payload is a little lighter compared to compliant JSONRPC protocol described below (i.e no `jsonrpc` version attribute, `method` become `m`, `params` become `p`, `result` become `r`)
* [Missive](https://github.com/StarryInternet/missive) package is used both server and client side to transform payloads into a binary format. Missive support zlib compression, but it's not used here and it's not recommended because of performance crumble. Missive is based on [fringe](https://github.com/StarryInternet/fringe), an extensible message framing over streams for nodejs.

### **JSONRPC**

The JSONRPC protocol has been chosen for interoperability.
```
REQUEST                                     | RESPONSE
------------------------------------------- | --------------------------------------
{                                           | {
    "jsonrpc":"2.0",                        |   "jsonrpc":"2.0"
    "method":"myMethod",                    |   "result":true,
    "params":["foo","bar"],                 |   "id":"operation-uniq-id"
    "id":"operation-uniq-id"                | }
}                                           |
```
* Raw and standard JSONRPC over TCP
* [jayson](https://github.com/tedeh/jayson) package is used server side. Actually only TCP transport is implemented, but HTTP(s) JSON API and websocket API may be implemented in the future.   

### **Dinary**

It's not a typo. Dinary use 2 Binary clients, one socket for requests, the other one for responses. This is the fastest protocol, the one by default.

The underling protocol is the same as the Binary one.

-----

## Server

### Installation

```
npm install -g sloki
```

### Run

```
sloki
```

## Help

```
sloki --help

=======================================================================
              Sloki - a NodeJS Server for LokyJS
=======================================================================
 Environment variable              Default
   SLOKI_TCP_BINARY_ENABLE         true
   SLOKI_TCP_BINARY_PORT           6370
   SLOKI_TCP_BINARY_HOST           localhost
   SLOKI_TCP_BINARY_MAX_CLIENTS    64
   SLOKI_TLS_BINARY_ENABLE         true
   SLOKI_TLS_BINARY_PORT           6371
   SLOKI_TLS_BINARY_HOST           localhost
   SLOKI_TLS_BINARY_MAX_CLIENTS    64

   SLOKI_TCP_JSONRPC_ENABLE        true
   SLOKI_TCP_JSONRPC_PORT          6372
   SLOKI_TCP_JSONRPC_HOST          localhost
   SLOKI_TCP_JSONRPC_MAX_CLIENTS   64
   SLOKI_TLS_JSONRPC_ENABLE        true
   SLOKI_TLS_JSONRPC_PORT          6373
   SLOKI_TLS_JSONRPC_HOST          localhost
   SLOKI_TLS_JSONRPC_MAX_CLIENTS   64

   SLOKI_TCP_DINARY_ENABLE         true
   SLOKI_TCP_DINARY_PORT           6374
   SLOKI_TCP_DINARY_HOST           localhost
   SLOKI_TCP_DINARY_MAX_CLIENTS    64
   SLOKI_TLS_DINARY_ENABLE         true
   SLOKI_TLS_DINARY_PORT           6375
   SLOKI_TLS_DINARY_HOST           localhost
   SLOKI_TLS_DINARY_MAX_CLIENTS    64

   SLOKI_DIR                       /home/franck/.sloki
   SLOKI_SHOW_OPS_INTERVAL         0
   SLOKI_GC_INTERVAL               3600000
   SLOKI_MEM_LIMIT                 26094 Mb
-----------------------------------------------------------------------
 Command Line Options              Default
   --tcp-binary-enable             true
   --tcp-binary-port               6370
   --tcp-binary-host               localhost
   --tcp-binary-max-clients        64
   --tls-binary-enable             true
   --tls-binary-port               6371
   --tls-binary-host               localhost
   --tls-binary-max-clients        64
   --tcp-jsonrpc-enable            true
   --tcp-jsonrpc-port              6372
   --tcp-jsonrpc-host              localhost
   --tcp-jsonrpc-max-clients       64
   --tls-jsonrpc-enable            true
   --tls-jsonrpc-port              6373
   --tls-jsonrpc-host              localhost
   --tls-jsonrpc-max-clients       64
   --tcp-dinary-enable             undefined
   --tcp-dinary-port               6374
   --tcp-dinary-host               localhost
   --tcp-dinary-max-clients        64
   --tls-dinary-enable             true
   --tls-dinary-port               6375
   --tls-dinary-host               localhost
   --tls-dinary-max-clients        64
   --dir                           /home/franck/.sloki
   --show-ops-interval             0
   --gc-interval                   3600000
   --mem-limit                     26094 Mb
-----------------------------------------------------------------------
Examples:
$ sloki     # will use defaults
$ sloki --tcp-binary-port=6370 --tcp-binary-host=localhost
```
-----

## Clients

### NodeJS
See https://github.com/sloki-project/sloki-node-client

The client will load every methods that sloki server have.



## Benchmarks

```
> npm run bench

#################################################################################
# Benchmark suite using sloki v0.0.8    Intel® Core™ i7-6820HQ 2.70Ghz
#################################################################################
# x64 | 8 CPU(s) | linux (4.4.0-43-Microsoft Linux) | node v11.10.0
#################################################################################
> client connected (binary)
> client connected (binarys)
> client connected (jsonrpc)
> client connected (jsonrpcs)
> client connected (dinary)
> client connected (dinarys)
>>>>> test insert#nocallback
> run insert#nocallback@binary
> run insert#nocallback@binarys
> run insert#nocallback@jsonrpc
> run insert#nocallback@jsonrpcs
> run insert#nocallback@dinary
> run insert#nocallback@dinarys
>>>>> test insert#callback.fullDocument
> run insert#callback.fullDocument@binary
> run insert#callback.fullDocument@binarys
> run insert#callback.fullDocument@jsonrpc
> run insert#callback.fullDocument@jsonrpcs
> run insert#callback.fullDocument@dinary
> run insert#callback.fullDocument@dinarys
>>>>> test insert#callback.sret.01
> run insert#callback.sret.01@binary
> run insert#callback.sret.01@binarys
> run insert#callback.sret.01@jsonrpc
> run insert#callback.sret.01@jsonrpcs
> run insert#callback.sret.01@dinary
> run insert#callback.sret.01@dinarys
>>>>> gc done (rss before 206 MB, after 194 MB)
> client disconnected (binary)
> client disconnected (binarys)
> client disconnected (jsonrpc)
> client disconnected (jsonrpcs)
> client disconnected (dinary)
> client disconnected (dinarys)

# --------------------------------------------------------------------------------
# Test                                     | Operations | ops/sec | exec time
# --------------------------------------------------------------------------------
# insert#nocallback@binary                 |      20000 |   13489 |    1.5s
# insert#nocallback@binarys                |      20000 |  406818 |    50ms
# insert#nocallback@jsonrpc                |      20000 |   21520 |   930ms
# insert#nocallback@jsonrpcs               |      20000 |  369549 |    55ms
# insert#nocallback@dinary                 |      20000 |   15348 |    1.3s
# insert#nocallback@dinarys                |      20000 | 1418742 |    15ms
# insert#callback.fullDocument@binary      |      20000 |   11745 |    1.7s
# insert#callback.fullDocument@binarys     |      20000 |   11899 |    1.7s
# insert#callback.fullDocument@jsonrpc     |      20000 |   19992 |      1s
# insert#callback.fullDocument@jsonrpcs    |      20000 |   21005 |   953ms
# insert#callback.fullDocument@dinary      |      20000 |   12324 |    1.6s
# insert#callback.fullDocument@dinarys     |      20000 |   33110 |   605ms
# insert#callback.sret.01@binary           |      20000 |   12189 |    1.6s
# insert#callback.sret.01@binarys          |      20000 |   34291 |   584ms
# insert#callback.sret.01@jsonrpc          |      20000 |   14420 |    1.4s
# insert#callback.sret.01@jsonrpcs         |      20000 |   21053 |   951ms
# insert#callback.sret.01@dinary           |      20000 |   12464 |    1.6s
# insert#callback.sret.01@dinarys          |      20000 |   32928 |   608ms
```

The winner is dinary protocol !

## Development status

Not usable yet.

-----


CHANGELOG.md
# Changelog

## [0.0.9](https://github.com/sloki-project/sloki/milestone/7) - 2019-03-02

### Added
* add "Dinary" protocol support [#33](https://github.com/sloki-project/sloki/issues/33)



## [0.0.8](https://github.com/sloki-project/sloki/milestone/6) - 2019-02-23

### Changed
* fix some require path for docker


## [0.0.7](https://github.com/sloki-project/sloki/milestone/5) - 2019-02-23

### Added
* add method `versions` [#19](https://github.com/sloki-project/sloki/issues/19)
* add alias property for all methods [#21](https://github.com/sloki-project/sloki/issues/21)
* add binary tcp protocol [#22](https://github.com/sloki-project/sloki/issues/22)
* add manual garbage collector [#23](https://github.com/sloki-project/sloki/issues/23) [#24](https://github.com/sloki-project/sloki/issues/24)
* TLS: generate self signed certificate when sloki start [#26](https://github.com/sloki-project/sloki/issues/26)
* TLS: implement both jsonrpc and binary server [#27](https://github.com/sloki-project/sloki/issues/27)

### Changed
* switch to named parameter in jsonrpc layer [#20](https://github.com/sloki-project/sloki/issues/20)
* switch to switch to loki-fs-structured-adapter by default [#32](https://github.com/sloki-project/sloki/issues/32)
* listen on multiple port per transport/protocol by default [#25](https://github.com/sloki-project/sloki/issues/25)


## [0.0.6](https://github.com/sloki-project/sloki/milestone/4) - 2019-02-09

### Added
* add method `wait` for testing purpose [(one sec)](https://github.com/sloki-project/sloki/commit/80c51ac81d18e18794f1782486aec9d8b4166c55)

### Changed
* fix db is not sticked to tcp session if db already exists [#17](https://github.com/sloki-project/sloki/issues/17)
* improve find filters [#18](https://github.com/sloki-project/sloki/issues/18)



## [0.0.5](https://github.com/sloki-project/sloki/milestone/3) - 2019-02-08

### Added
* add method collection/`update` [#5](https://github.com/sloki-project/sloki/issues/5)
* add method collection/`find` [#3](https://github.com/sloki-project/sloki/issues/3)
* add method collection/`remove` [#4](https://github.com/sloki-project/sloki/issues/4)
* add method collection/`get` [#2](https://github.com/sloki-project/sloki/issues/2)

### Changed
* fix client destroy connexion when client end [#15](https://github.com/sloki-project/sloki/issues/15)
* fix insert return behavior [#9](https://github.com/sloki-project/sloki/issues/9)
* improve code: merge jsonrpc/lokijs layers [#14](https://github.com/sloki-project/sloki/issues/14)
* improve method collection/insert options [#11](https://github.com/sloki-project/sloki/issues/11)
* improve sanity check [#10](https://github.com/sloki-project/sloki/issues/10)


## [0.0.4](https://github.com/sloki-project/sloki/milestone/2) - 2019-02-07

### Added
* add JSONRPC over tcp layer (jayson)
* add method server/`clients`
* add method server/`methods`
* add method server/`maxClients`
* add method server/`quit`
* add method server/`shutdown`
* add method database/`db`
* add method database/`loadDatabase`
* add method database/`listDatabases`
* add method database/`saveDatabase`
* add method database/`addCollection`
* add method database/`listCollections`
* add method collection/`get`
* add method collection/`insert`
* add tests



## 0.0.0

* Start


src/config.js
const argv = require('minimist')(process.argv.slice(2));
const path = require('path');
const homedir = require('os').homedir();
const totalmem = require('os').totalmem();

const defaults = {

    // manual garbage collector, use --expose-gc to enable it
    // default is 1 hour
    GC_INTERVAL: 1000*60*60,
    MEM_LIMIT:Math.round((totalmem-(totalmem*0.20))/1024/1024),   // in Mb
    MEM_LIMIT_REACHED:false,

    // default database dir is in the home of the user
    SLOKI_DIR:path.resolve(homedir+'/.sloki'),

    DATABASES_DEFAULT_OPTIONS :  {
        //serializationMethod:'pretty',
        autoload:true,
        autosave:true,
        autosaveInterval:1000*60*5,
        throttledSaves:false
    },

    RE_COLLETION_NAME: '^[a-z0-9\-\.\_]{1,50}$',
    RE_DATABASE_NAME: '^[a-z0-9\-\.\_]{1,50}$',

    // http://jsonrpc.org/spec.html#error_object
    ERROR_CODE_PARAMETER: -32602,
    ERROR_CODE_INTERNAL: -32603,

    // listeners
    TCP_BINARY_ENABLE:true,
    TCP_BINARY_PORT:6370,
    TCP_BINARY_HOST:'localhost',
    TCP_BINARY_MAX_CLIENTS:64,

    TLS_BINARY_ENABLE:true,
    TLS_BINARY_PORT:6371,
    TLS_BINARY_HOST:'localhost',
    TLS_BINARY_MAX_CLIENTS:64,

    TCP_JSONRPC_ENABLE:true,
    TCP_JSONRPC_PORT:6372,
    TCP_JSONRPC_HOST:'localhost',
    TCP_JSONRPC_MAX_CLIENTS:64,

    TLS_JSONRPC_ENABLE:true,
    TLS_JSONRPC_PORT:6373,
    TLS_JSONRPC_HOST:'localhost',
    TLS_JSONRPC_MAX_CLIENTS:64,

    TCP_DINARY_ENABLE:true,
    TCP_DINARY_PORT:6374,
    TCP_DINARY_HOST:'localhost',
    TCP_DINARY_MAX_CLIENTS:64,

    TLS_DINARY_ENABLE:true,
    TLS_DINARY_PORT:6375,
    TLS_DINARY_HOST:'localhost',
    TLS_DINARY_MAX_CLIENTS:64,

    // Show number of operation per interval
    SHOW_OPS_INTERVAL:0,
};

if (process.env.DOCKER) {
    const ip = '0.0.0.0';
    defaults.TCP_BINARY_HOST = ip;
    defaults.TLS_BINARY_HOST = ip;
    defaults.TCP_JSONRPC_HOST = ip;
    defaults.TLS_JSONRPC_HOST = ip;
    defaults.TCP_DINARY_HOST = ip;
    defaults.TLS_DINARY_HOST = ip;
}

const c = JSON.parse(JSON.stringify(defaults));

const transports = ['TCP', 'TLS'];
const protocols = ['BINARY', 'JSONRPC', 'DINARY'];
const opts = ['ENABLE', 'PORT', 'HOST', 'MAX_CLIENTS'];

/*******************************************
 * command line help with default values
 *******************************************/

if (argv.help) {
    console.log();
    console.log('=======================================================================');
    console.log('              Sloki - a NodeJS Server for LokyJS                       ');
    console.log('=======================================================================');
    console.log(' Environment variable              Default                             ');
    console.log(`   SLOKI_TCP_BINARY_ENABLE         ${c.TCP_BINARY_ENABLE}              `);
    console.log(`   SLOKI_TCP_BINARY_PORT           ${c.TCP_BINARY_PORT}                `);
    console.log(`   SLOKI_TCP_BINARY_HOST           ${c.TCP_BINARY_HOST}                `);
    console.log(`   SLOKI_TCP_BINARY_MAX_CLIENTS    ${c.TCP_BINARY_MAX_CLIENTS}         `);
    console.log(`   SLOKI_TLS_BINARY_ENABLE         ${c.TLS_BINARY_ENABLE}              `);
    console.log(`   SLOKI_TLS_BINARY_PORT           ${c.TLS_BINARY_PORT}                `);
    console.log(`   SLOKI_TLS_BINARY_HOST           ${c.TLS_BINARY_HOST}                `);
    console.log(`   SLOKI_TLS_BINARY_MAX_CLIENTS    ${c.TLS_BINARY_MAX_CLIENTS}         `);
    console.log();
    console.log(`   SLOKI_TCP_JSONRPC_ENABLE        ${c.TCP_JSONRPC_ENABLE}             `);
    console.log(`   SLOKI_TCP_JSONRPC_PORT          ${c.TCP_JSONRPC_PORT}               `);
    console.log(`   SLOKI_TCP_JSONRPC_HOST          ${c.TCP_JSONRPC_HOST}               `);
    console.log(`   SLOKI_TCP_JSONRPC_MAX_CLIENTS   ${c.TCP_JSONRPC_MAX_CLIENTS}        `);
    console.log(`   SLOKI_TLS_JSONRPC_ENABLE        ${c.TLS_JSONRPC_ENABLE}             `);
    console.log(`   SLOKI_TLS_JSONRPC_PORT          ${c.TLS_JSONRPC_PORT}               `);
    console.log(`   SLOKI_TLS_JSONRPC_HOST          ${c.TLS_JSONRPC_HOST}               `);
    console.log(`   SLOKI_TLS_JSONRPC_MAX_CLIENTS   ${c.TLS_JSONRPC_MAX_CLIENTS}        `);
    console.log();
    console.log(`   SLOKI_TCP_DINARY_ENABLE         ${c.TCP_DINARY_ENABLE}              `);
    console.log(`   SLOKI_TCP_DINARY_PORT           ${c.TCP_DINARY_PORT}                `);
    console.log(`   SLOKI_TCP_DINARY_HOST           ${c.TCP_DINARY_HOST}                `);
    console.log(`   SLOKI_TCP_DINARY_MAX_CLIENTS    ${c.TCP_DINARY_MAX_CLIENTS}         `);
    console.log(`   SLOKI_TLS_DINARY_ENABLE         ${c.TLS_DINARY_ENABLE}              `);
    console.log(`   SLOKI_TLS_DINARY_PORT           ${c.TLS_DINARY_PORT}                `);
    console.log(`   SLOKI_TLS_DINARY_HOST           ${c.TLS_DINARY_HOST}                `);
    console.log(`   SLOKI_TLS_DINARY_MAX_CLIENTS    ${c.TLS_DINARY_MAX_CLIENTS}         `);
    console.log();
    console.log(`   SLOKI_DIR                       ${c.SLOKI_DIR}                      `);
    console.log(`   SLOKI_SHOW_OPS_INTERVAL         ${c.SHOW_OPS_INTERVAL}              `);
    console.log(`   SLOKI_GC_INTERVAL               ${c.GC_INTERVAL}                    `);
    console.log(`   SLOKI_MEM_LIMIT                 ${c.MEM_LIMIT} Mb                   `);
    console.log('-----------------------------------------------------------------------');
    console.log(' Command Line Options              Default                             ');
    console.log(`   --tcp-binary-enable             ${c.TCP_BINARY_ENABLE}              `);
    console.log(`   --tcp-binary-port               ${c.TCP_BINARY_PORT}                `);
    console.log(`   --tcp-binary-host               ${c.TCP_BINARY_HOST}                `);
    console.log(`   --tcp-binary-max-clients        ${c.TCP_BINARY_MAX_CLIENTS}         `);
    console.log(`   --tls-binary-enable             ${c.TLS_BINARY_ENABLE}              `);
    console.log(`   --tls-binary-port               ${c.TLS_BINARY_PORT}                `);
    console.log(`   --tls-binary-host               ${c.TLS_BINARY_HOST}                `);
    console.log(`   --tls-binary-max-clients        ${c.TLS_BINARY_MAX_CLIENTS}         `);
    console.log(`   --tcp-jsonrpc-enable            ${c.TCP_JSONRPC_ENABLE}             `);
    console.log(`   --tcp-jsonrpc-port              ${c.TCP_JSONRPC_PORT}               `);
    console.log(`   --tcp-jsonrpc-host              ${c.TCP_JSONRPC_HOST}               `);
    console.log(`   --tcp-jsonrpc-max-clients       ${c.TCP_JSONRPC_MAX_CLIENTS}        `);
    console.log(`   --tls-jsonrpc-enable            ${c.TLS_JSONRPC_ENABLE}             `);
    console.log(`   --tls-jsonrpc-port              ${c.TLS_JSONRPC_PORT}               `);
    console.log(`   --tls-jsonrpc-host              ${c.TLS_JSONRPC_HOST}               `);
    console.log(`   --tls-jsonrpc-max-clients       ${c.TLS_JSONRPC_MAX_CLIENTS}        `);
    console.log(`   --tcp-dinary-enable             ${c.TCP_DINARY_PORTINARY_ENABLE}              `);
    console.log(`   --tcp-dinary-port               ${c.TCP_DINARY_PORT}                `);
    console.log(`   --tcp-dinary-host               ${c.TCP_DINARY_HOST}                `);
    console.log(`   --tcp-dinary-max-clients        ${c.TCP_DINARY_MAX_CLIENTS}         `);
    console.log(`   --tls-dinary-enable             ${c.TLS_DINARY_ENABLE}              `);
    console.log(`   --tls-dinary-port               ${c.TLS_DINARY_PORT}                `);
    console.log(`   --tls-dinary-host               ${c.TLS_DINARY_HOST}                `);
    console.log(`   --tls-dinary-max-clients        ${c.TLS_DINARY_MAX_CLIENTS}         `);
    console.log(`   --dir                           ${c.SLOKI_DIR}                      `);
    console.log(`   --show-ops-interval             ${c.SHOW_OPS_INTERVAL}              `);
    console.log(`   --gc-interval                   ${c.GC_INTERVAL}                    `);
    console.log(`   --mem-limit                     ${c.MEM_LIMIT} Mb                   `);
    console.log('-----------------------------------------------------------------------');
    console.log('Examples:                                                              ');
    console.log('$ sloki     # will use defaults                                        ');
    console.log('$ sloki --tcp-binary-port=6370 --tcp-binary-host=localhost             ');
    console.log();
    process.exit();
}

/***********************************************************************
 * c override: env var, then cmd line option
 ***********************************************************************/

for (const t of transports) {
    for (const e of protocols) {
        for (const o of opts) {
            const v = `${t}_${e}_${o}`;
            const l = v.toLowerCase().replace(/\_/g, '-');
            if (process.env[`SLOKI_${v}`] != undefined) {
                c[v] = process.env[`SLOKI_${v}`];
            } else if (argv[l] != undefined) {
                c[v] = argv[l];
            }
        }
    }
}

c.SLOKI_DIR = process.env.SLOKI_DIR || argv['dir'] || c.SLOKI_DIR;
c.SHOW_OPS_INTERVAL = parseInt(process.env.SLOKI_SHOW_OPS_INTERVAL || argv['show-ops-interval'] || c.SHOW_OPS_INTERVAL);
c.GC_INTERVAL = parseInt(process.env.SLOKI_GC_INTERVAL || argv['gc-interval'] || c.GC_INTERVAL);
c.MEM_LIMIT = parseInt(process.env.SLOKI_MEM_LIMIT || argv['mem-limit'] || c.MEM_LIMIT);


/********************************
 * integrity checks
 ********************************/
function exitError(msg) {
    console.log(`Error: ${msg}`);
    process.exit(-1);
}

// check enable/disable transport/protocol

for (const t of transports) {
    for (const e of protocols) {
        const v = `${t}_${e}_ENABLE`;
        const o = '--'+v.toLowerCase().replace(/\_/g, '-');
        if (c[v] === 'true') c[v] = true;
        if (c[v] === 'false') c[v] = false;
        if (typeof c[v] != 'boolean') {
            exitError(`Error: SLOKI_${v} or ${o} value should be a boolean (true/false)`);
        }
    }
}

// check ports

const maxPort = 65534;

for (const t of transports) {
    for (const e of protocols) {
        const v = `${t}_${e}_PORT`;
        const o = '--'+v.toLowerCase().replace(/\_/g, '-');
        c[v] = parseInt(c[v]);
        if (isNaN(c[v]) || c[v]<1 || c[v]>maxPort) {
            exitError(`Error: SLOKI_${v} or ${o} value should be a number between 1 and ${maxPort}`);
        }
    }
}

// check max clients

const maxClients = 1024;

for (const t of transports) {
    for (const e of protocols) {
        const v = `${t}_${e}_MAX_CLIENTS`;
        const o = '--'+v.toLowerCase().replace(/\_/g, '-');
        c[v] = parseInt(c[v]);
        if ((isNaN(c[v]) || !c[v]) || (c[v]<1 || c[v]>maxClients)) {
            exitError(`Error: SLOKI_${v} or ${o} value should be a number between 0 and ${maxClients}`);
        }
    }
}

// others

c.SLOKI_DIR = path.resolve(c.SLOKI_DIR);
c.SLOKI_DIR_DBS = path.resolve(c.SLOKI_DIR+'/dbs');

if (isNaN(c.GC_INTERVAL) || !c.GC_INTERVAL) {
    exitError('SLOKI_GC_INTERVAL or --gc-interval value should be a number (milliseconds)');
}

if (isNaN(c.MEM_LIMIT) || !c.MEM_LIMIT) {
    exitError('SLOKI_MEM_LIMIT or --mem-limit value should be in MegaBytes, and > 0');
}

c.getMaxClients = function(protocol) {
    switch (protocol) {
    case 'tcp': return c.TCP_BINARY_MAX_CLIENTS;
    case 'tls': return c.TLS_BINARY_MAX_CLIENTS;
    case 'binary': return c.TCP_BINARY_MAX_CLIENTS;
    case 'binarys': return c.TLS_BINARY_MAX_CLIENTS;
    case 'jsonrpc': return c.TCP_JSONRPC_MAX_CLIENTS;
    case 'jsonrpcs': return c.TLS_JSONRPC_MAX_CLIENTS;
    case 'dinary': return c.TCP_DINARY_MAX_CLIENTS;
    case 'dinarys': return c.TLS_DINARY_MAX_CLIENTS;
    default:throw new Error(`unknow protocol '${protocol}'`);
    }
};

c.setMaxClients = function(protocol, max) {
    switch (protocol) {
    case 'tcp': c.TCP_BINARY_MAX_CLIENTS = max;break;
    case 'tls': c.TLS_BINARY_MAX_CLIENTS = max;break;
    case 'binary': c.TCP_BINARY_MAX_CLIENTS = max;break;
    case 'binarys': c.TLS_BINARY_MAX_CLIENTS = max;break;
    case 'jsonrpc': c.TCP_JSONRPC_MAX_CLIENTS = max;break;
    case 'jsonrpcs': c.TLS_JSONRPC_MAX_CLIENTS = max;break;
    case 'dinary': c.TCP_DINARY_MAX_CLIENTS = max;break;
    case 'dinarys': c.TLS_DINARY_MAX_CLIENTS = max;break;
    default:throw new Error(`unknow protocol '${protocol}'`);
    }
};

module.exports = c;
